File: tests/07_phases_ompss.dg/c/success_task_final_01.c
/*--------------------------------------------------------------------
  (C) Copyright 2006-2012 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/



/*
<testinfo>
test_generator=config/mercurium-ompss
test_compile_fail_nanos6_mercurium=yes
test_compile_fail_nanos6_imcc=yes
</testinfo>
*/

#include<assert.h>

int omp_in_final()
{
    assert(0);
    return -1;
}

int f(int n) {
    if (n <= 0) return 0;

    int res = 0;
    #pragma omp task shared(res)
    {
        res += f(n-1);
        res += omp_in_final() ? 1 : 2;
    }

    return res;
}

int main(int argc, char*argv[])
{
    int res = 0;
    #pragma omp task shared(res) final(1)
        res = f(10);
    #pragma omp taskwait
    assert(res == 10);
}

File: tests/07_phases_ompss.dg/fortran/success_task_final_01.f90
! <testinfo>
! test_generator=config/mercurium-ompss
! test_compile_fail_nanos6_mercurium=yes
! test_compile_fail_nanos6_imfc=yes
! </testinfo>
FUNCTION OMP_IN_FINAL() RESULT(X)
    IMPLICIT NONE
    LOGICAL X
    STOP -1
    X = .FALSE.
END FUNCTION OMP_IN_FINAL

RECURSIVE FUNCTION F(N) RESULT(RES)
    IMPLICIT NONE
    INTEGER :: N, RES
    LOGICAL :: X
    LOGICAL, EXTERNAL  :: OMP_IN_FINAL
    RES = 0
    IF (N > 0) THEN
    !$OMP TASK SHARED(RES)
        RES = F(N-1)
        IF(OMP_IN_FINAL()) THEN
            RES = RES + 1
        ELSE
            RES = RES + 2
        ENDIF
    !$OMP END TASK
    ENDIF
END FUNCTION F

PROGRAM P
    IMPLICIT NONE
    INTEGER :: RES
    INTEGER, EXTERNAL :: F
    RES = 0

    !$OMP TASK SHARED(RES) FINAL(.TRUE.)
        RES = F(10)
    !$OMP END TASK
    !$OMP TASKWAIT

    IF (RES /= 10) STOP -2
END PROGRAM P
